urlb44e9843ba_QuasimodoNZ_SWEN222_Game_Project_tgz-pJ8-Game_World_GameWorldLevelBuilderJ8/src/Game_World/GameWorld.java

```diff
@@ -0,0 +1,476 @@
+	 * @param String name of Player
+	 * @param GameObject Food
+	 */
+	public void eatFood(String name, GameObject object) {
+		for (Avatar a : avatarLocations.keySet()) {
+			if (a.getName().equals(name)) {
+				a.eatFood((Food) object);
+			}
+		}
+	}
+
+	/**
+	 * Enables an avatar to pick up a GameObject.
+	 * @param String name of Player
+	 * @param GameObject
+	 */
+	public void pickupItem(String name, GameObject object) {
+		for (Avatar a : avatarLocations.keySet()) {
+			if (a.getName().equals(name)) {
+				a.pickUpObject(object);
+			}
+		}
+	}
+
+	/**
+	 * Enables an avatar to drop a GameObject.
+	 * @param String name of Player
+	 * @param GameObject
+	 */
+	public void dropItem(String name, GameObject object) {
+		for (Avatar a : avatarLocations.keySet()) {
+			if (a.getName().equals(name)) {
+				a.dropObject(object);
+			}
+		}
+	}
+
+	/**
+	 * Enables an avatar to move from one Location to another in the GameWorld.
+	 * @param String name of Player
+	 * @param GameObject
+	 */
+	public synchronized void moveThroughDoor(String name, ExternalDoor door) {
+		boolean locked = door.isLocked();
+		String endLocationName = door.getEndLocationName();
+		Point endLocationPoint = door.getEndLocationPoint();
+		int direction = door.getDirection();
+
+		for (Avatar a : avatarLocations.keySet()) {
+			if (a.getName().equals(name)) {
+				if (!locked && door.avatarIsClose(a)) {// if the door is not locked and the
+					// Avatar is about 1cm from the door.
+					GameWorld avatarWorld = a.getWorld();
+					Location avatarLocation = a.getWorld().getAvatarLocations().get(a);
+					avatarWorld.getAvatarLocations().remove(a);
+					avatarWorld.getAvatarLocations().put(a,avatarWorld.getLocations().get(endLocationName));
+					// move the Avatar from the old location's contents to the new location's contents.
+					avatarLocation.getAllObjects().remove(a);
+					avatarWorld.getLocations().get(endLocationName).getAllObjects().add(a);
+					a.setCurrentLocation(this.getLocations().get(endLocationName));
+					a.setLocationName(endLocationName);
+					// Reset Avatar's boundingbox to a a point in the new location
+					Rectangle r = a.getBoundingBox();
+					r.setLocation(endLocationPoint);
+					a.setLocationName(endLocationName);
+					//System.out.println(a.getLocationName());
+					a.setBoundingBox(r);
+					a.setDirection(direction);
+
+				}
+				return;
+			}
+		}
+	}
+
+	/**
+	 * Enables an avatar to close an internal door.
+	 * @param String name of Player
+	 * @param GameObject
+	 */
+	public void closeDoor(String name, InternalDoor door) {
+		Avatar a = null;
+		for(Avatar b: avatars){
+			if(b.getName().equals(name)){
+				a = b;
+			}
+		}
+		if(a!= null){
+			String locName = a.getLocationName();
+			String doorName = door.getName();
+			Location l = locations.get(locName);
+			for(GameObject g: l.getAllObjects()){
+				if(g.getName().equals(doorName)){
+					InternalDoor d = (InternalDoor) g;
+					d.setLocked(true);
+				}
+			}
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 320:336, 346:362.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.